File: leetcode/python/0002. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        place = 1
        res = 0
        while l1 != None or l2 != None:
            if l1 is None:
                s1 = 0
            else:
                s1 = l1.val * place
                l1 = l1.next
            
            if l2 is None:
                s2 = 0
            else:
                s2 = l2.val * place
                l2 = l2.next
            
            res += (s1 + s2)
            place *= 10
            
        res = str(res)
        node = None
        root = None
        for d in reversed(res):
            newnode = ListNode(d)
            if not node:
                node = newnode
                root = node
            else:
                node.next = newnode
                node = newnode
        return root

File: leetcode/java/0002. Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // Dummy head to simplify edge cases (no need to handle first node separately)
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;

        int carry = 0; // Carry from previous addition

        // Loop until both lists are fully processed
        while (l1 != null || l2 != null) {
            // Get values from current nodes (0 if null)
            int x = (l1 != null) ? l1.val : 0;
            int y = (l2 != null) ? l2.val : 0;

            // Sum current digits + carry
            int sum = x + y + carry;

            // Update carry for next iteration
            carry = sum / 10;

            // Create new node with the unit digit
            curr.next = new ListNode(sum % 10);
            curr = curr.next;

            // Move to next nodes if available
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }

        // If carry remains, append it as a new node
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }

        return dummy.next; // First node after dummy is the result
    }
}
